util/circular_buffer.c

```diff
@@ -141,105 +147,266 @@ int32_t cb_remove(struct circular_buffer *desc)
+ *  - -EBUSY    - Asynchronous transaction already started
  */
-int32_t cb_read(struct circular_buffer *desc, void *data,
-		uint32_t nb_elements)
+int32_t cb_prepare_async_write(struct circular_buffer *desc,
+			       uint32_t size_to_write,
+			       void **write_buff,
+			       uint32_t *size_avilable)
 {
-	uint32_t	to_read_size;
-	uint32_t	size_to_end;
+	return cb_prepare_async_operation(desc, size_to_write, write_buff,
+					  size_avilable, 0);
+}
 
-	if (!desc || !nb_elements)
-		return FAILURE;
+/**
+ * @brief Prepare asynchronous read
+ *
+ * Get the inside raw buffer to be used in DMA transactions.
+ *
+ * @param desc - Circular buffer reference
+ * @param size_to_read - Number of bytes needed to write to the buffer.
+ * @param read_buff - Address where to store the buffer where data will be read.
+ * @param size_avilable - min(size_to_read, size until end of allocated buffer)
+ * @return
+ *  - \ref SUCCESS   - No errors
+ *  - -EAGAIN   - No data available at this moment
+ *  - -EINVAL   - Wrong parameters used
+ *  - -EBUSY    - Asynchronous transaction already started
+ *  - -EOVERRUN - An overrun occurred and some data have been overwritten
+ */
+int32_t cb_prepare_async_read(struct circular_buffer *desc,
+			      uint32_t size_to_read,
+			      void **read_buff,
+			      uint32_t *size_avilable)
+{
+	return cb_prepare_async_operation(desc, size_to_read, read_buff,
+					  size_avilable, 1);
+}
 
-	/* Check for overflow */
-	if (nb_elements > desc->nb_elem)
-		return FAILURE;
-	desc->nb_elem -= nb_elements;
-
-	to_read_size = nb_elements * desc->elem_size;
-	size_to_end = desc->size - (desc->read_ptr - desc->buff);
-	if (to_read_size > size_to_end) {
-		memcpy(data, desc->read_ptr, size_to_end);
-		to_read_size -= size_to_end;
-		memcpy((int8_t *)data + size_to_end, desc->buff, to_read_size);
-		desc->read_ptr = desc->buff + to_read_size;
-	} else {
-		memcpy(data, desc->read_ptr, to_read_size);
-		desc->read_ptr += to_read_size;
-	}
+/**
+ * \defgroup end_async_group End Ashyncronous functions
+ * @brief End asynchronous transaction
+ *
+ * @param desc - Circular buffer reference
+ * @return
+ *  - \ref SUCCESS   - No errors
+ *  - \ref FAILURE   - Asynchronous transaction not started
+ *  - -EINVAL        - Wrong parameters used
+ * @{
+ */
+int32_t cb_end_async_write(struct circular_buffer *desc)
+{
+	return cb_end_async_operation(desc, 0);
+}
 
-	return SUCCESS;
+int32_t cb_end_async_read(struct circular_buffer *desc)
+{
+
+	return cb_end_async_operation(desc, 1);
 }
+/** @} */
 
+/**
+ * @brief Write data to the buffer (Blocking)
+ * @param desc - Circular buffer reference
+ * @param data - Buffer from where data is copied to the circular buffer
+ * @param size - Size to write
+ * @return
+ *  - \ref SUCCESS - No errors
+ *  - -EINVAL      - Wrong parameters used
+ */
+int32_t cb_write(struct circular_buffer *desc, const void *data, uint32_t size)
+{
+	return cb_operation(desc, (void *)data, size, 0);
```

Comment:
Are these wrappers `cb_write` and `cb_read` really necessary?

Can't we just define
`#define WRITE 0`
and
`#define READ 1`

And use directly `cb_operation(desc, data, size, WRITE)` and `cb_operation(desc, data, size, READ)` 

Just an opinion.